actors/builtin/miner/miner_actor.go

```diff
@@ -565,65 +538,96 @@ func (a Actor) ConfirmSectorProofsValid(rt Runtime, params *builtin.ConfirmSecto
+		// TODO: we should batch these calls...
+		// https://github.com/filecoin-project/specs-actors/issues/474
+		_, code := rt.Send(
+			builtin.StorageMarketActorAddr,
+			builtin.MethodsMarket.ActivateDeals,
+			&market.ActivateDealsParams{
+				DealIDs:      precommit.Info.DealIDs,
+				SectorExpiry: precommit.Info.Expiration,
+			},
+			abi.NewTokenAmount(0),
+		)
+		builtin.RequireSuccess(rt, code, "failed to activate deals")
 
-		// Add sector and pledge lock-up to miner state
-		// TODO: do this all at once after the loop
-		// https://github.com/filecoin-project/specs-actors/issues/476
-		newlyVestedAmount := rt.State().Transaction(&st, func() interface{} {
-			newlyVestedFund, err := st.UnlockVestedFunds(store, rt.CurrEpoch())
-			if err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to vest new funds: %s", err)
-			}
+		preCommits = append(preCommits, precommit)
+	}
 
-			// Unlock deposit for successful proof, make it available for lock-up as initial pledge.
-			st.AddPreCommitDeposit(precommit.PreCommitDeposit.Neg())
-			st.AddInitialPledgeRequirement(initialPledge)
+	// This transaction should replace the one inside the loop above.
+	// Migrate state mutations into here.
+	totalPledge := big.Zero()
+	newSectors := make([]*SectorOnChainInfo, 0)
+	var info *MinerInfo
+	newlyVestedAmount := rt.State().Transaction(&st, func() interface{} {
+		err := scheduleReplaceSectorsExpiration(rt, &st, store, replaceSectors)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to replace sector expirations")
 
-			// Lock up initial pledge for new sector.
-			availableBalance := st.GetAvailableBalance(rt.CurrentBalance())
-			if availableBalance.LessThan(initialPledge) {
-				rt.Abortf(exitcode.ErrInsufficientFunds, "insufficient funds for initial pledge requirement %s, available: %s", initialPledge, availableBalance)
-			}
-			if err := st.AddLockedFunds(store, rt.CurrEpoch(), initialPledge, &PledgeVestingSpec); err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to add pledge: %v", err)
+		info = getMinerInfo(rt, &st)
+
+		for _, precommit := range preCommits {
+			// initial pledge is precommit deposit
+			initialPledge := precommit.PreCommitDeposit
+			totalPledge = big.Add(totalPledge, initialPledge)
+			newSectorInfo := SectorOnChainInfo{
+				SectorNumber:       precommit.Info.SectorNumber,
+				SealProof:          precommit.Info.SealProof,
+				SealedCID:          precommit.Info.SealedCID,
+				DealIDs:            precommit.Info.DealIDs,
+				Expiration:         precommit.Info.Expiration,
+				Activation:         precommit.PreCommitEpoch,
+				DealWeight:         precommit.DealWeight,
+				VerifiedDealWeight: precommit.VerifiedDealWeight,
+				InitialPledge:      initialPledge,
 			}
-			st.AssertBalanceInvariants(rt.CurrentBalance())
+			newSectors = append(newSectors, &newSectorInfo)
 
 			if err := st.PutSector(store, &newSectorInfo); err != nil {
```

Comment:
@anorth it wasn't totally clear but from conversation in #476 it might be that you want to go further and take these four sector state operations out of this loop creating batch calls (`PutSectors`, `DeletePrecommittedSectors`, etc) on the state.  Let me know if that's the case.

Reply:
I do also want to do that, but defer it to a follow up. It's probably worth a new issue for "batch miner state mutations for sectors, precommits"